[exporter/elasticsearch] add support for `batcher` config

axw opened the pull request:
**Description:**

Add opt-in support for the experimental batch sender (https://github.com/open-telemetry/opentelemetry-collector/issues/8122). When opting into this functionality the exporter's `Consume*` methods will make synchronous bulk requests to Elasticsearch, without additional batching/buffering in the exporter.

By default the exporter continues to use its own buffering, which supports thresholds for time, number of documents, and size of encoded documents in bytes. The batch sender does not currently support a bytes-based threshold, and is experimental, hence why we are not yet making it the default for the Elasticsearch exporter.

This PR is based on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32632, but made to be non-breaking.

**Link to tracking Issue:**

#32377

**Testing:**

Added unit and integration tests.

Manually tested with Elasticsearch, using the following config to highlight that client metadata can now flow through all the way:

```yaml
receivers:
  otlp:
    protocols:
      http:
        endpoint: 0.0.0.0:4318
        include_metadata: true

exporters:
  elasticsearch:
    endpoint: "http://localhost:9200"
    auth:
      authenticator: headers_setter
    batcher:
      enabled: false

extensions:
  headers_setter:
    headers:
      - action: insert
        key: Authorization
        from_context: authorization

service:
  extensions: [headers_setter]
  pipelines:
    traces:
      receivers: [otlp]
      processors: []
      exporters: [elasticsearch]
```

I have Elasticsearch running locally, with an "admin" user with the password "changeme". Sending OTLP/HTTP to the collector with `telemetrygen traces --otlp-http --otlp-insecure http://localhost:4318 --otlp-header "Authorization=\"Basic YWRtaW46Y2hhbmdlbWU=\""`, I observe the following:
 - Without the `batcher` config, the exporter fails to index data into Elasticsearch due to an auth error. That's because the exporter is buffering and dropping the context with client metadata, so there's no Authorization header attached to the requests going out.
 - With `batcher: {enabled: true}`, same behaviour as above. Unlike the [`batch` processor](https://github.com/open-telemetry/opentelemetry-collector/tree/main/processor/batchprocessor), the batch sender does not maintain client metadata.
 - With `batcher: {enabled: false}`, the exporter successfully indexes data into Elasticsearch.

**Documentation:**

Updated the README.

Commit:
[exporter/elasticsearch] add sync bulk indexer

Add opt-in support for the experimental batcher helper -
https://github.com/open-telemetry/opentelemetry-collector/issues/8122

When opting into this functionality the exporter's Consume*
methods will make synchronous bulk requests to Elasticsearch,
without additional batching/buffering in the exporter.

Commit:
Fix lint errors

Commit:
American spelling

carsonip commented on exporter/elasticsearchexporter/README.md:
```diff
@@ -81,6 +81,32 @@ All other defaults are as defined by [confighttp].
+> The `batcher` config is experimental and may change without notice.
+
+The Elasticsearch exporter supports the [common `batcher` settings](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterbatcher/config.go).
+The `batcher` config is currently disabled by default.
```
A little bit of ambiguity here but I don't know how to make it clearer. Strictly speaking, it is unset by default.

axw replied:
I've reworded the paragraph below to make it clearer that the exporter does its own buffering/batching by default, let me know if it looks good to you.

carsonip commented on exporter/elasticsearchexporter/factory.go:
```diff
@@ -164,9 +168,29 @@ func createTracesExporter(ctx context.Context,
+			MinSizeConfig: cfg.Batcher.MinSizeConfig,
+			MaxSizeConfig: cfg.Batcher.MaxSizeConfig,
+		}
+		opts = append(opts, exporterhelper.WithBatcher(batcherConfig))
```
```suggestion
		opts = append(opts, exporterhelper.WithBatcher(batcherConfig))
		opts = append(opts, exporterhelper.WithTimeout(exporterhelper.TimeoutSettings{Timeout: 0})) // effectively disable timeout_sender because timeout is enforced in bulk indexer
```

axw replied:
That's only true for the async bulk indexer though - the sync bulk indexer does not add a timeout. Should it?

carsonip replied:
Here are 2 questions:
- Should sync bulk indexer add a timeout? In my PR, yes, but I see why you implemented it without a timeout. If we decide that each retry in sync bulk indexer can take inf time, we should state it in readme that sync bulk indexer ignores `timeout` (confighttp::timeout). Alternatively, timeout defaults to 0 for sync and another default for async.
- Setting `timeout_sender: 0` is a must, as this is sync, and default timeout_sender is 5s (questionable imo). This will potentially override any timeout we enforce / not enforce in sync bulk indexer.

axw replied:
I had a bit of a brainfart, and I conflated flush timeout with flush _interval_, hence why I didn't honour it for sync. I'm going to update the sync bulk indexer to honour the flush timeout, after which it makes sense to disable `timeout_sender`.

IIUC we still want `timeout_sender` for the async bulk indexer, since otherwise there's no timeout applied to the channel send in `Add`.

carsonip commented on exporter/elasticsearchexporter/integrationtest/exporter_test.go:
```diff
@@ -20,27 +20,41 @@ func TestExporter(t *testing.T) {
 			mockESFailure    bool
 		}{
 			{name: "basic"},
+			{name: "batcher_enabled", batcherEnabled: ptrTo(true)},
```
non blocker: should we also test with 2 cases, `restartCollector: true` and `mockESFailure: true`?

axw replied:
I've added test cases for `mockESFailure: true`, but not `restartCollector: true`. IIANM the latter would require adding persistent queuing to the integration tests, which I'd rather leave to a follow up if it's OK with you.

carsonip reviewed (commented):
thanks! looks great at a high level

Commit:
Reword README about flush being the default

Commit:
Add tests for batcher + intermittent ES failures

axw commented:
Apparently the backoff function is stateful, as can be seen from the data races in the tests. Working on a fix.

axw committed:
Update exporter/elasticsearchexporter/factory.go

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

Commit:
Fixes

- Add flush timeout to sync bulk indexer, disable timeout_sender.
- Create a new backoff function for each call to Flush in the sync
  bulk indexer, since the function is stateful and not safe for
  concurrent calls to Flush.
